plutus-core/untyped-plutus-core/src/Language/UntypedPlutusCore/Evaluation/Machine/Cek/ExBudgetMode.hs

```diff
@@ -0,0 +1,140 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+{-# LANGUAGE StrictData            #-}
+
+module Language.UntypedPlutusCore.Evaluation.Machine.Cek.ExBudgetMode
+    ( CekExTally (..)
+    , ExBudgetMode (..)
+    , CountingSt (..)
+    , TallyingSt (..)
+    , RestrictingSt (..)
+    , Hashable
+    , counting
+    , tallying
+    , restricting
+    , restrictingEnormous
+    )
+where
+
+import           PlutusPrelude
+
+import           Language.UntypedPlutusCore.Evaluation.Machine.Cek.Internal
+
+import           Language.PlutusCore.Evaluation.Machine.ExBudget
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Machine.Exception
+
+import           Control.Lens                                               (ifoldMap)
+import           Control.Monad.Except
+import           Control.Monad.State.Strict
+import           Data.HashMap.Monoidal                                      as HashMap
+import           Data.Hashable                                              (Hashable)
+import           Data.List                                                  (intersperse)
+import qualified Data.Map.Strict                                            as Map
+import           Data.Semigroup.Generic
+import           Data.Text.Prettyprint.Doc
+import           Text.PrettyBy                                              (IgnorePrettyConfig (..))
+
+-- | A budgeting mode to execute an evaluator in.
+data ExBudgetMode st uni fun = ExBudgetMode
+    { _exBudgetModeSpender :: CekBudgetSpender st uni fun  -- ^ A spending function.
```

Comment:
This all feels like it ought to be something nice and standard: we're basically just doing a monadic accumulating fold. Maybe something like `FoldM` (https://hackage.haskell.org/package/foldl-1.4.11/docs/Control-Foldl.html#t:FoldM)? I don't know if it would make any difference, it just feels a little itchy.